### PR TITLE
[JENKINS-40825] move transients into DecoratedLauncher

### PR DESCRIPTION
**Pull request by marvinthepa**
as `decorate` will be called multiple times, and
we don't want to share these transients among calls -
especially as they will not be properly cleaned up.

Makes JENKINS-40825 much less severe, as
"Pipe not connected" now most of the time seems to
come from [DurableTaskStep][1], which will just retry
instead of failing the job.

More importantly, it fixes the resource leak mentioned in
[this comment](https://issues.jenkins-ci.org/browse/JENKINS-40825?focusedCommentId=308905&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-308905) and the following comments.

Note: this is just a first stab and probably needs polishing, especially
the new `ContainerExecWatcher.close` implementation, which
might not even be necessary.

[1]: https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/ae18393/src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java#L332

### CODE REVIEW

**Comment by marvinthepa** [comment]
Whoops, forgot the tests. Will fix them right now.

**Comment by marvinthepa** [comment]
Tests pass on my system, but I needed a small "hack" to make the slaves running (on virtualbox) be able to connect to my Jenkins. Will open an extra pull request for that.

**Comment by marvinthepa** [comment]
Looks like this is not enough. Please don't merge yet.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 1]
```diff
@@ -100,198 +97,222 @@ public ContainerExecDecorator(KubernetesClient client, String podName, String co
...
-    public Launcher decorate(final Launcher launcher, final Node node) {
-        return new Launcher.DecoratedLauncher(launcher) {
-            @Override
-            public Proc launch(ProcStarter starter) throws IOException {
-                if (!waitUntilContainerIsReady()) {
-                    throw new IOException("Failed to execute shell script inside container " +
-                            "[" + containerName + "] of pod [" + podName + "]." +
-                            " Timed out waiting for container to become ready!");
-                }
+    private class ClosableLauncher extends Launcher.DecoratedLauncher implements Closeable {
```
Comment (iocanel) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Do we really need to create a new class? Can't we just have the exiting class add things to a list of closeables?

**Comment by marvinthepa** [comment]
@iocanel:

First of all, thanks for the review.

The new class was for the implementation of the `close` method, but I now realize that it could have been an List<Closeable> instead, so I could have done without it, right.

Anyway, I realized that this approach is not enough anyway - Even with this pull request, we are overwriting `started`, `finished`, and `alive` from different threads, so I am currently trying to move them into local variables instead of members.

Closing this pull request now, expect a new one tomorrow.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 1 above)
Comment (marvinthepa) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Right, see https://github.com/jenkinsci/kubernetes-plugin/pull/180#issuecomment-319718931

**Comment by iocanel** [comment]
@marvinthepa: If by "this is not enough" you mean that there are still stuck threads, I noticed in the client that the InputStreamPumper.close() method didn't properly broke the pumping loop (in some cases), causing the thread to remain stuck. It seems to work somehow nicer, with this: https://github.com/fabric8io/kubernetes-client/commit/61ae6ed33fbe19dd49d3b34d206257e77d39355e which is now available in: https://github.com/fabric8io/kubernetes-client/releases/tag/v2.5.8 (which is in central). So you might want to try this version out.

**Comment by marvinthepa** [comment]
I actually meant that we might still end up not even calling the `close` method of some of the resources that we use - so even if the `close` method works correctly, we might leave resources around.

See [996952cad3](https://github.com/marvinthepa/kubernetes-plugin/commit/996952cad3851a431efaed77511d80d264233368?w=1)
resp. [this branch](https://github.com/marvinthepa/kubernetes-plugin/commits/bugfix/JENKINS-40825-broken-pipe-now-really) for what I think works better.

I will try to verify a few of my assumptions (especially about `proc` being killed) and might polish it a bit more, and then create a new pull request.

**Comment by iocanel** [comment]
@marvinthepa: If we know for sure that proc is being killed by the shell step, then things get simplified a lot, as it frees us completely from having to babysit and hold refs of the watch and proc.

**Comment by marvinthepa** [comment]
Proc is not what is being killed by the shell step, but actually `ContainerExecDecorator`. So we need to actually keep a reference around for every proc that we started. I don't really like this, but I don't see an alternative.

**Comment by michaelajr** [comment]
Am I understanding this correctly - with this pull request, the "Pipe not connected" issue will be seen (logged) but will no longer "freeze" the build for 5 mins and then fail it?

**Comment by marvinthepa** [comment]
@michaelajr:
Not exactly, but I am working on it. See [this branch](https://github.com/marvinthepa/kubernetes-plugin/commits/bugfix/JENKINS-40825-broken-pipe-now-really) for progress, as the approach in this branch was flawed.

If you have a test system, feel free to build from that branch and give it a spin, I would be interested to get feedback if it works for you.

You still might have issues when it takes you more than 10 seconds to start processes in kubernetes (i.e. likely when your kubernetes is overwhelmed), but the general situation should improve.

I managed to get rid of the pipe not connected errors, but I still need to get the `proc` cleanup right, that is why I haven't created a pull request yet.

**Comment by michaelajr** [comment]
@marvinthepa OK. We'll take a look. Thanks!

**Comment by marvinthepa** [comment]
Please see #182 instead.

**Pull request closed without merging**